python/mlc_chat/model/llama/llama_model.py

```diff
@@ -199,44 +225,73 @@ def __init__(self, config: LlamaConfig):
         self.norm = nn.RMSNorm(config.hidden_size, -1, config.rms_norm_eps, bias=False)
         self.tensor_parallel_shards = config.tensor_parallel_shards
 
-    def forward(self, inputs: Tensor, total_seq_len: tir.Var, attention_mask: Tensor):
-        if self.tensor_parallel_shards > 1:
-            inputs = op.ccl_broadcast_from_worker0(inputs)
-        hidden_states = self.embed_tokens(inputs)
-        for layer in self.layers:
-            hidden_states = layer(hidden_states, attention_mask, total_seq_len)
+    def forward(
+        self, input_ids: Optional[Tensor], input_embeds: Optional[Tensor], attn_args: AttentionArgs
+    ):
+        if input_ids is not None:
+            assert input_embeds is None
+            if self.tensor_parallel_shards > 1:
+                input_ids = op.ccl_broadcast_from_worker0(input_ids)
+            hidden_states = self.embed_tokens(input_ids)
+        else:
+            assert input_embeds is not None
+            if self.tensor_parallel_shards > 1:
+                input_embeds = op.ccl_broadcast_from_worker0(input_embeds)
+            hidden_states = input_embeds
```

Comment:
this is suboptimal to have a huge `if` and a couple of optional arguments used like tagged union. perhaps we could keep a few different `forward` methods in this case as well